### ISSUE TITLE
How do I reproduce the result of the FLERT model?

### ISSUE THREAD

**Issue by xuuuluuu** [flairNLP/flair#2200]
Hi, 

Do you have a description of how to reproduce the results in the FLERT paper?

Thanks a lot.

**Comment by alanakbik** [comment]
The scripts used to train each model can be found on the HF hub. For instance the English model was trained with this script: https://huggingface.co/flair/ner-english-large (scroll down to the "Training: Script to train this model" part). You can find the same scripts for the other "large" models on the HF hub.

**Comment by xuuuluuu** [comment]
Hi, thanks for the reply.

I have followed the same configuration in the link, and the following are the logs. The micro F1 on the CoNLL2003 English dataset is  93.20.

Is it possible to suggest where the issue is?



ps. I also tried with BERT-large-cased,  the micro F1 is 92.38.




          (23): RobertaLayer(
            (attention): RobertaAttention(
              (self): RobertaSelfAttention(
                (query): Linear(in_features=1024, out_features=1024, bias=True)
                (key): Linear(in_features=1024, out_features=1024, bias=True)
                (value): Linear(in_features=1024, out_features=1024, bias=True)
                (dropout): Dropout(p=0.1, inplace=False)
              )
              (output): RobertaSelfOutput(
                (dense): Linear(in_features=1024, out_features=1024, bias=True)
                (LayerNorm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
                (dropout): Dropout(p=0.1, inplace=False)
              )
            )
            (intermediate): RobertaIntermediate(
              (dense): Linear(in_features=1024, out_features=4096, bias=True)
            )
            (output): RobertaOutput(
              (dense): Linear(in_features=4096, out_features=1024, bias=True)
              (LayerNorm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
          )
        )
      )
      (pooler): RobertaPooler(
        (dense): Linear(in_features=1024, out_features=1024, bias=True)
        (activation): Tanh()
      )
    )
  )
  (word_dropout): WordDropout(p=0.05)
  (locked_dropout): LockedDropout(p=0.5)
  (linear): Linear(in_features=1024, out_features=20, bias=True)
  (beta): 1.0
  (weights): None
  (weight_tensor) None
)"
2021-04-08 18:17:08,929 ----------------------------------------------------------------------------------------------------
2021-04-08 18:17:08,929 Corpus: "Corpus: 14987 train + 3466 dev + 3684 test sentences"
2021-04-08 18:17:08,929 ----------------------------------------------------------------------------------------------------
2021-04-08 18:17:08,929 Parameters:
2021-04-08 18:17:08,929  - learning_rate: "5e-06"
2021-04-08 18:17:08,929  - mini_batch_size: "4"
2021-04-08 18:17:08,929  - patience: "3"
2021-04-08 18:17:08,929  - anneal_factor: "0.5"
2021-04-08 18:17:08,930  - max_epochs: "20"
2021-04-08 18:17:08,930  - shuffle: "True"
2021-04-08 18:17:08,930  - train_with_dev: "False"
2021-04-08 18:17:08,930  - batch_growth_annealing: "False"
2021-04-08 18:17:08,930 ----------------------------------------------------------------------------------------------------
2021-04-08 18:17:08,930 Model training base path: "resources/taggers/ner-english-large"
2021-04-08 18:17:08,930 ----------------------------------------------------------------------------------------------------

...


2021-04-09 03:27:15,589 Testing using best model ...
2021-04-09 03:29:28,119 0.9249	0.9393	0.9320
2021-04-09 03:29:28,119 
Results:
- F1-score (micro) 0.9320
- F1-score (macro) 0.9175

By class:
LOC        tp: 1559 - fp: 80 - fn: 109 - precision: 0.9512 - recall: 0.9347 - f1-score: 0.9428
MISC       tp: 604 - fp: 157 - fn: 98 - precision: 0.7937 - recall: 0.8604 - f1-score: 0.8257
ORG        tp: 1556 - fp: 161 - fn: 105 - precision: 0.9062 - recall: 0.9368 - f1-score: 0.9213
PER        tp: 1586 - fp: 33 - fn: 31 - precision: 0.9796 - recall: 0.9808 - f1-score: 0.9802
2021-04-09 03:29:28,119 ----------------------------------------------------------------------------------------------------

**Comment by alanakbik** [comment]
Hm, I ran a similar script on master branch last week and got the same result as in the paper (averaged over 3 runs). For the paper models, we didn't use context dropout and annealed against train loss, so perhaps you could train like this: 

```python
import torch

# 1. get the corpus
from flair.datasets import CONLL_03

corpus = CONLL_03()

# 2. what tag do we want to predict?
tag_type = 'ner'

# 3. make the tag dictionary from the corpus
tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)

# 4. initialize fine-tuneable transformer embeddings WITH document context
from flair.embeddings import TransformerWordEmbeddings

embeddings = TransformerWordEmbeddings(
    model='xlm-roberta-large',
    layers="-1",
    subtoken_pooling="first",
    fine_tune=True,
    use_context=True,
    context_dropout=0.,
)

# 5. initialize bare-bones sequence tagger (no CRF, no RNN, no reprojection)
from flair.models import SequenceTagger

tagger = SequenceTagger(
    hidden_size=256,
    embeddings=embeddings,
    tag_dictionary=tag_dictionary,
    tag_type='ner',
    use_crf=False,
    use_rnn=False,
    reproject_embeddings=False,
)

# 6. initialize trainer with AdamW optimizer
from flair.trainers import ModelTrainer

trainer = ModelTrainer(tagger, corpus, optimizer=torch.optim.AdamW)

# 7. run training with XLM parameters (20 epochs, small LR)
from torch.optim.lr_scheduler import OneCycleLR

trainer.train('resources/taggers/ner-english-large',
              learning_rate=5.0e-6,
              mini_batch_size=4,
              mini_batch_chunk_size=1,
              train_with_dev=True,
              max_epochs=20,
              scheduler=OneCycleLR,
              embeddings_storage_mode='none',
              weight_decay=0.,
              )
```

**Comment by alanakbik** [comment]
BTW are you on the latest Flair release or master branch?

**Comment by xuuuluuu** [comment]
Thank you! 

I will try the new configs.  I used the master branch.

**Comment by alanakbik** [comment]
Best use the last release! The master branch is undergoing constant change at the moment as we are preparing some larger changes.

**Comment by xuuuluuu** [comment]
With the latest release and the attached configs:

train_with_dev = True    -->    93.71
train_with_dev = False   -->    93.56

It shows that removing the context dropout helps. Are there any other configs I can tweak a little bit although the performance difference is minor?


By setting use_context = 512   --> 93.80

I am wondering about the differences between the FLERT (when set context length to 512) and the BERT (maximal document context). (except for the model size)
